File: Concurrency/producer-consumer.py
import os
import asyncio
from asyncio import Queue
import time
import tracemalloc

tracemalloc.start()  # start tracing memory allocations

os.environ["PYTHONASYNCIODEBUG"] = "1"

producers = 1000
consumers = 1000
capacity = 10


async def producer(queue, id):
    await queue.put(id)
    print(f"Producer {id} produced")


async def consumer(queue, id):
    value = await queue.get()
    print(f"Consumer {id} consumed value {value}")


async def main():
    queue = Queue(capacity)
    tasks = []

    # Launch producer tasks
    for i in range(producers):
        tasks.append(asyncio.create_task(producer(queue, i)))

    # Launch consumer tasks
    for i in range(consumers):
        tasks.append(asyncio.create_task(consumer(queue, i)))

    start_time = time.monotonic()
    await asyncio.gather(*tasks)
    elapsed_time = time.monotonic() - start_time
    print(f"Total execution time: {elapsed_time:.2f} seconds")

    print("---")


asyncio.run(main())

current, peak = tracemalloc.get_traced_memory()
print(f"Peak memory usage: {peak / 1024 / 1024:.2f} MB")
tracemalloc.stop()  # stop tracing memory allocations

File: Concurrency/producer-consumer.go
package main

import (
	"fmt"
	"runtime"
	"sync"
	"time"
)

const (
	producers = 1000
	consumers = 1000
	capacity  = 10
)

func main() {
	startTime := time.Now()
	// Set the number of CPU cores used by the program to 1
	var m runtime.MemStats

	runtime.ReadMemStats(&m) // read current memory statistics
	runtime.GOMAXPROCS(1)

	var wg sync.WaitGroup
	wg.Add(producers + consumers)

	buffer := make(chan int, capacity)

	// Launch producer Goroutines
	for i := 0; i < producers; i++ {
		go func(id int) {
			defer wg.Done()
			buffer <- id
			fmt.Printf("Producer %d produced\n", id)
		}(i)
	}

	// Launch consumer Goroutines
	for i := 0; i < consumers; i++ {
		go func(id int) {
			defer wg.Done()
			value := <-buffer
			fmt.Printf("Consumer %d consumed value %d\n", id, value)
		}(i)
	}

	wg.Wait()

	fmt.Println("Done")
	var memStats runtime.MemStats
	runtime.ReadMemStats(&memStats)

	peakMemoryUsage := float64(memStats.HeapSys) / (1024 * 1024)
	fmt.Printf("Peak memory usage: %.2f MB\n", peakMemoryUsage)

	elapsedTime := time.Since(startTime)
	fmt.Printf("Total execution time: %.6f seconds\n", elapsedTime.Seconds())
}
